FWIO/RNTuple/plugins/RNTupleSource.cc

```diff
@@ -0,0 +1,328 @@
+#include "FWCore/Framework/interface/EventPrincipal.h"
+#include "FWCore/Framework/interface/LuminosityBlockPrincipal.h"
+#include "FWCore/Framework/interface/RunPrincipal.h"
+#include "FWCore/Framework/interface/SharedResourcesRegistry.h"
+#include "FWCore/Framework/interface/InputSourceDescription.h"
+#include "FWCore/Framework/interface/PreallocationConfiguration.h"
+#include "FWCore/Framework/interface/FileBlock.h"
+#include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+
+#include "FWCore/Utilities/interface/propagate_const.h"
+
+#include "RNTupleInputFile.h"
+#include "RNTupleDelayedReader.h"
+#include <memory>
+
+namespace edm {
+
+  namespace {
+    class ProvenanceReader : public edm::ProvenanceReaderBase {
+    public:
+      ProvenanceReader(edm::input::DataProductsRNTuple* iRNTuple,
+                       std::vector<ParentageID> const* iParentageIDLookup,
+                       std::vector<int>::const_iterator iStreamToEntryBegin,
+                       SharedResourcesAcquirer* iAcquirer,
+                       std::recursive_mutex* iMutex)
+          : rntuple_(iRNTuple),
+            acquirer_(iAcquirer),
+            mutex_(iMutex),
+            parentageIDLookup_(iParentageIDLookup),
+            entryBegin_(iStreamToEntryBegin),
+            descriptor_(rntuple_->findDescriptorID("EventProductProvenance")) {}
+
+      std::set<ProductProvenance> readProvenance(unsigned int transitionIndex) const final;
+      void readProvenanceAsync(WaitingTaskHolder task,
+                               ModuleCallingContext const* moduleCallingContext,
+                               unsigned int transitionIndex,
+                               std::atomic<const std::set<ProductProvenance>*>& writeTo) const noexcept final;
+
+    public:
+      edm::input::DataProductsRNTuple* rntuple_;
+      SharedResourcesAcquirer* acquirer_;
+      std::recursive_mutex* mutex_;
+      std::vector<ParentageID> const* parentageIDLookup_;
+      std::vector<int>::const_iterator entryBegin_;
+      ROOT::Experimental::DescriptorId_t descriptor_;
+    };
+
+    std::set<ProductProvenance> ProvenanceReader::readProvenance(unsigned int transitionIndex) const {
+      std::vector<StoredProductProvenance> provVec;
+      {
+        std::shared_ptr<std::vector<StoredProductProvenance>> provPtr{&provVec, [](auto) {}};
+        std::lock_guard<std::recursive_mutex> guard(*mutex_);
+        auto v = rntuple_->viewFor(descriptor_, provPtr);
+        auto entry = *(entryBegin_ + transitionIndex);
+        v(entry);
+      }
+
+      std::set<ProductProvenance> retValue;
+      for (auto const& prov : provVec) {
+        if (prov.parentageIDIndex_ >= parentageIDLookup_->size()) {
+          throw edm::Exception(errors::LogicError)
+              << "ReducedProvenanceReader::ReadProvenance\n"
```

Comment:
```suggestion
              << "RNTuple ProvenanceReader::readProvenance\n"
```
or something similar

Reply:
done